examples/simple.rs

```diff
@@ -1,67 +1,120 @@
-use hyper::server::conn::AddrStream;
-use hyper::service::{make_service_fn, service_fn};
-use hyper::{Body, Request, Response, Server, StatusCode};
+use std::convert::Infallible;
+use std::io;
+use std::net::{IpAddr, SocketAddr};
+use std::time::Duration;
+
+use http_body_util::combinators::UnsyncBoxBody;
+use http_body_util::{BodyExt, Empty, Full};
+use hyper::body::{Bytes, Incoming};
+use hyper::server::conn::http1;
+use hyper::service::service_fn;
+use hyper::{Request, Response, StatusCode};
+use hyper_util::rt::{TokioExecutor, TokioIo, TokioTimer};
+use tokio::net::TcpListener;
+
 use hyper_reverse_proxy::ReverseProxy;
-use hyper_trust_dns::{RustlsHttpsConnector, TrustDnsResolver};
-use std::net::IpAddr;
-use std::{convert::Infallible, net::SocketAddr};
+use hyper_rustls::{ConfigBuilderExt, HttpsConnector};
+use hyper_util::client::legacy::connect::HttpConnector;
+
+type Connector = HttpsConnector<HttpConnector>;
+type ResponseBody = UnsyncBoxBody<Bytes, std::io::Error>;
 
 lazy_static::lazy_static! {
-    static ref  PROXY_CLIENT: ReverseProxy<RustlsHttpsConnector> = {
-        ReverseProxy::new(
-            hyper::Client::builder().build::<_, hyper::Body>(TrustDnsResolver::default().into_rustls_webpki_https_connector()),
-        )
+    static ref PROXY_CLIENT: ReverseProxy<Connector> = {
+      let connector: Connector = Connector::builder()
+          .with_tls_config(
+              rustls::ClientConfig::builder()
+                  .with_native_roots()
+                  .expect("with_native_roots")
+                  .with_no_client_auth(),
+          )
+          .https_or_http()
+          .enable_http1()
+          .build();
+      ReverseProxy::new(
+          hyper_util::client::legacy::Builder::new(TokioExecutor::new())
+              .pool_idle_timeout(Duration::from_secs(3))
+              .pool_timer(TokioTimer::new())
+              .build::<_, Incoming>(connector),
+      )
     };
 }
 
-fn debug_request(req: &Request<Body>) -> Result<Response<Body>, Infallible> {
-    let body_str = format!("{:?}", req);
-    Ok(Response::new(Body::from(body_str)))
-}
-
-async fn handle(client_ip: IpAddr, req: Request<Body>) -> Result<Response<Body>, Infallible> {
-    if req.uri().path().starts_with("/target/first") {
+async fn handle(
+    client_ip: IpAddr,
+    req: Request<Incoming>,
+) -> Result<Response<ResponseBody>, Infallible> {
+    let host = req.headers().get("host").and_then(|v| v.to_str().ok());
+    if host.is_some_and(|host| host.starts_with("service1.localhost")) {
```

Comment:
The motivation for using the host header over the path is because the path is forwarded to the upstream server, and most servers will return a useless 404 for the path `/target/first`. Meanwhile, most servers will ignore what is in the host header, and most browsers let you use subdomains on localhost now, so http://service1.localhost should work for most cases out of the box. Using `starts_with` since the port number will be at the end!